### PR TITLE
[MNG-7470] mvn 3.9+ IT that uses wagon (default) and native transport

### PR DESCRIPTION
**Pull request by cstamas**
An IT for Maven 3.9+ that "exercise" both resolver transport: wagon and native (so we have 2 tests both forcing transport with `-Dmaven.resolver.transport` one wagon (which is default, but setting is there to be explicit) and once native.

It builds a project that needs a plugin (to pull it will use set transport), and that plugin will use resolver as well to resolve set artifact. If any of these fails, maven project fails.

Current maven 3.9.x build will FAIL due NoSuchMethod error, as expected.

With PR https://github.com/apache/maven/pull/732

```
Bootstrap.Bootstrap()........................................ OK (3.1 s)
mng7470 ResolverTransport.ResolverTransportNative().......... 3332 [main] INFO org.eclipse.jetty.server.Server - jetty-9.0.4.v20130625
3397 [main] INFO org.eclipse.jetty.server.ServerConnector - Started ServerConnector@6032d904{HTTP/1.1}{0.0.0.0:34895}
OK (1.7 s)
mng7470 ResolverTransport.ResolverTransportWagon()........... OK (1.6 s)
```

### CODE REVIEW

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            HashMap<String, String> properties = new HashMap<>();
+            properties.put( "@port@", Integer.toString( server.port() ) );
+            verifier.filterFile( "settings-template.xml", "settings.xml", "UTF-8", properties );
+
+            verifier = newVerifier( new File( testDir, "wagon" ).getAbsolutePath() );
+            verifier.deleteDirectory( "target" );
+            verifier.deleteArtifacts( "org.apache.maven.its.resolver-transport.wagon" );
+            verifier.addCliOption( "-X" );
+            verifier.addCliOption("-s" );
+            verifier.addCliOption(new File( testDir, "settings.xml" ).getAbsolutePath() );
```
Comment (michael-o) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:42]: space missing

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            HashMap<String, String> properties = new HashMap<>();
+            properties.put( "@port@", Integer.toString( server.port() ) );
+            verifier.filterFile( "settings-template.xml", "settings.xml", "UTF-8", properties );
+
+            verifier = newVerifier( new File( testDir, "native" ).getAbsolutePath() );
+            verifier.deleteDirectory( "target" );
+            verifier.deleteArtifacts( "org.apache.maven.its.resolver-transport.native" );
+            verifier.addCliOption( "-X" );
+            verifier.addCliOption("-s" );
+            verifier.addCliOption(new File( testDir, "settings.xml" ).getAbsolutePath() );
```
Comment (michael-o) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:77]: space missing

**Review by michael-o — CHANGES_REQUESTED** [review]
Where is the verification that the transport actually works, i.e., the right one is picked?

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java (hunk 1 above)
Comment (cstamas) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:42]: > Where is the verification that the transport actually works, i.e., the right one is picked?

Maven build fails if project deps cannot be resolved... is that what you asking?

**Comment by cstamas** [comment]
> Where is the verification that the transport actually works, i.e., the right one is picked?

Maven build fails if project deps cannot be resolved... is that what you asking?

**Comment by michael-o** [comment]
> > Where is the verification that the transport actually works, i.e., the right one is picked?
> 
> Maven build fails if project deps cannot be resolved... is that what you asking?

Yes, this is what I will look for. Can you add this as a comment to the tests. It is not obvious from the code itself.

**Comment by cstamas** [comment]
> Yes, this is what I will look for. Can you add this as a comment to the tests. It is not obvious from the code itself.

Done.

**Comment by michael-o** [comment]
Testing...

**Comment by cstamas** [comment]
Do not merge pls, I'll do myself once done

**Comment by michael-o** [comment]
[INFO] Downloading from maven-core-it: http://localhost:53770/org/apache/maven/its/resolver-transport/wagon/dependency/1.0/dependency-1.0.pom
[WARNING] Checksum validation failed, expected '97641443fee307b6c3593080be865df89ace3d86' (REMOTE_EXTERNAL) but is actually 'cd16a85341375c7f9dda5edd75dedb83e213e425' from maven-core-it for http://localhost:53770/org/apache/maven/its/resolver-transport/wagon/dependency/1.0/dependency-1.0.pom

**Comment by michael-o** [comment]
On
```
Java version: 1.8.0_322, vendor: OpenJDK BSD Porting Team, runtime: /usr/local/openjdk8/jre
Default locale: de_DE, platform encoding: UTF-8
OS name: "freebsd", version: "12.3-stable", arch: "amd64", family: "unix"
```

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            verifier.filterFile( "settings-template.xml", "settings.xml", "UTF-8", properties );
+
+            verifier = newVerifier( new File( testDir, "wagon" ).getAbsolutePath() );
+            verifier.deleteDirectory( "target" );
+            verifier.deleteArtifacts( "org.apache.maven.its.resolver-transport.wagon" );
+            verifier.addCliOption( "-X" );
+            verifier.addCliOption("-s" );
+            verifier.addCliOption( new File( testDir, "settings.xml" ).getAbsolutePath() );
+            verifier.addCliOption( "-Pmaven-core-it-repo" );
+            verifier.addCliOption( "-Dmaven.resolver.transport=wagon" );
```
Comment (slawekjaranowski) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:44]: I can remove it and test still pass ...

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            verifier.filterFile( "settings-template.xml", "settings.xml", "UTF-8", properties );
+
+            verifier = newVerifier( new File( testDir, "native" ).getAbsolutePath() );
+            verifier.deleteDirectory( "target" );
+            verifier.deleteArtifacts( "org.apache.maven.its.resolver-transport.native" );
+            verifier.addCliOption( "-X" );
+            verifier.addCliOption("-s" );
+            verifier.addCliOption( new File( testDir, "settings.xml" ).getAbsolutePath() );
+            verifier.addCliOption( "-Pmaven-core-it-repo" );
+            verifier.addCliOption( "-Dmaven.resolver.transport=native" );
```
Comment (slawekjaranowski) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:82]: I can remove it and test still pass ...

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java (hunk 3 above)
Comment (slawekjaranowski) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:44]: only one line ...
```
 verifier.addCliOption( "-Dmaven.resolver.transport=wagon" );
```

**Comment by cstamas** [comment]
@michael-o @slawekjaranowski so pushed all I wanted, and locally tested, updated description.

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java (hunk 4 above)
Comment (cstamas) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:82]: sure, as it fallbacks to wagon, that's not the point here

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java (hunk 3 above)
Comment (cstamas) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:44]: sure, as wagon is the default, but is here only to be explicit

**Comment by michael-o** [comment]
Will test again...

**Comment by slawekjaranowski** [comment]
We should also check what happens if wrong transport will be provided, like `maven.resolver.transport=xxx`
We will be sure that property is used at all

**Comment by cstamas** [comment]
> Is it ok?

That should not happen, especially as now assertion to debug log is added. Did you return the removed line?

**Comment by cstamas** [comment]
> We should also check what happens if wrong transport will be provided, like `maven.resolver.transport=xxx` We will be sure that property is used at all

IMHO, this is not a UT, this IT merely ensures that selected transport IS used and WORKS.

**Comment by cstamas** [comment]
@slawekjaranowski cool, that's expected with latest maven 3.9.x build, as it uses resolver 1.8.0 but has no fix in https://github.com/apache/maven/pull/732 Now, to verify, build locally that PR and run ITs with it, and they should all pass OK

**Comment by slawekjaranowski** [comment]
Both tests use the same project directory .... pleas add

```
verifier.setLogFileName( ... );
```

without it we have logs in file `log.txt` which content logs from last executed test

**Comment by cstamas** [comment]
> Both tests use the same project directory .... pleas add
> 
> ```
> verifier.setLogFileName( ... );
> ```
> 
> without it we have logs in file `log.txt` which content logs from last executed test

fixed

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    private void verifyLogHasLine( final Verifier verifier, final String logSnippet )
+            throws VerificationException
+    {
+        List<String> lines = verifier.loadFile( verifier.getBasedir(), verifier.getLogFileName(), false );
+
+        for ( String line : lines )
+        {
+            if ( !line.contains( logSnippet ) )
```
Comment (slawekjaranowski) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:68]: should be
```
if ( line.contains( logSnippet ) )
```

Comment (cstamas) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java]: fixed

**Review by slawekjaranowski — APPROVED** [review]
- build pass
- assertions works

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            HashMap<String, String> properties = new HashMap<>();
+            properties.put( "@port@", Integer.toString( server.port() ) );
+            verifier.filterFile( "settings-template.xml", "settings.xml", "UTF-8", properties );
+
+            verifier = newVerifier( new File( testDir, "project" ).getAbsolutePath() );
+            verifier.setLogFileName( transport + "-transport.log" );
+            verifier.deleteDirectory( "target" );
+            verifier.deleteArtifacts( "org.apache.maven.its.resolver-transport" );
+            verifier.addCliOption( "-X" );
+            verifier.addCliOption("-s" );
```
Comment (michael-o) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:42]: space

**Review by michael-o — CHANGES_REQUESTED** [review]
I am quite surprised that Jetty output is visible, MNG-5771 has a Jetty-based test as well, but no output is seen. Any idea? I do not see any diff in the source code.

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            verifyLogHasLine( verifier, logSnippet );
+            verifier.resetStreams();
+        }
+        finally
+        {
+            server.stop();
+        }
+    }
+
+    private void verifyLogHasLine( final Verifier verifier, final String logSnippet )
```
[the remark this answers is not in the document]
Comment (cstamas) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:61]: Agreed

[the remark this answers is not in the document]
Comment (slawekjaranowski) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng7470ResolverTransportTest.java:61]: There is `verifier.verifyTextInLog(...)`

**Comment by michael-o** [comment]
Still testing...

**Comment by michael-o** [comment]
Works for me now except for the JEtty log output which I don't understand....

**Comment by cstamas** [comment]
https://github.com/apache/maven/pull/732 is merged, so am merging this one as well (ongoing branches not having merged PR will fail here)

**Pull request merged**